packages/gatsby/src/utils/find-page-by-path.ts

```diff
@@ -0,0 +1,67 @@
+import { IGatsbyPage, IGatsbyState } from "../redux/types"
+import { match } from "@reach/router/lib/utils"
+
+export function findPageByPath(
+  state: IGatsbyState,
+  path: string,
+  fallbackTo404: boolean = false
+): IGatsbyPage | undefined {
+  const { pages } = state
+
+  // first check by exact path
+  let page = pages.get(path)
+  if (page) {
+    return page
+  }
+
+  if (path === ``) {
+    // from my tests I never was able to make request with
+    // completely empty pathname, but just for the sake
+    // of completeness - try available alternative
+    page = pages.get(`/`)
+    if (page) {
+      return page
+    }
+  }
+  // Gatsby doesn't allow for page path to be empty string,
+  // so skipping trying to get page for "" path if we can't
+  // find page for `/`
+  else if (path !== `/`) {
```

Comment:
So the case when the path equals to `/` is handled with the "exact match" codepath at lines 12-15?

Reply:
Yes, from what I tested there is only one way for index route - and this is is just `/` case. We can't create pages with empty path because it doesn't pass validation we have in `createPage` action creator (`""` is falsy)  - https://github.com/gatsbyjs/gatsby/blob/051e1aa5c168fa75f04ea496c1e99802ae8a5511/packages/gatsby/src/redux/actions/public.js#L176-L191